[WIP] kubeadm and add-on docs

lukemarsden opened the pull request:
DO NOT MERGE

This is a WIP.

This PR introduces:
* a new "quickstart" on how to use kubeadm to easily install Kubernetes
* a new "add-ons" page, intended to be a user-friendly central place for folks to go to look at the list of add-ons they can install on their cluster

Currently I've only added Weaveworks add-ons to the latter, but we should encourage other add-ons to be added to this list.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1265)
<!-- Reviewable:end -->

lavalamp commented:
please squash this :)

lukemarsden commented:
@lavalamp squashed :)

caseydavenport commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+
+### Joining your nodes
+
+The nodes are where your containers (your workload) will run.
```
s/containers/Pods ?

s/workload/workloads

lukemarsden replied:
fixed, thanks!

caseydavenport commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+
+### Installing a pod network
+
+You must install a pod network add-on so that your pods can communicate with eachother on different hosts.
```
s/eachother/each other

caseydavenport commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+
+1. The cluster we create here won't have cloud-provider integrations, so for example won't work with (for example) [Load Balancers](/docs/user-guide/load-balancer/) (LBs) or [Persistent Volumes](/docs/user-guide/persistent-volumes/walkthrough/) (PVs).
+   
+   Instead we will use the [NodePort feature of services](/docs/user-guide/services/) to demonstrate exposing the sample application on the internet.
```
This line is in the future tense, but the user has already _used_ the NodePort in the past :)

lukemarsden replied:
Will fix

lukemarsden replied:
fixed

caseydavenport commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+   Instead we will use the [NodePort feature of services](/docs/user-guide/services/) to demonstrate exposing the sample application on the internet.
+   To easily obtain a cluster which works with LBs and PVs Kubernetes, try [the "hello world" GKE tutorial](/docs/hellonode) or [one of the other cloud-specific installation tutorials](/docs/getting-started-guides/).
+1. The cluster we create here will have a single master, with a single `etcd` database running on it.
+   Adding HA support (multiple `etcd` servers, multiple API servers, etc) is still a work-in-progress.
```
Adding HA support to the `kubeadm` tool is still a work-in-progress (Kubernetes supports it in the general case)

lukemarsden replied:
fixed, thanks!

caseydavenport commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+   
+   Instead we will use the [NodePort feature of services](/docs/user-guide/services/) to demonstrate exposing the sample application on the internet.
+   To easily obtain a cluster which works with LBs and PVs Kubernetes, try [the "hello world" GKE tutorial](/docs/hellonode) or [one of the other cloud-specific installation tutorials](/docs/getting-started-guides/).
+1. The cluster we create here will have a single master, with a single `etcd` database running on it.
```
s/will have/has

Looks like this section was intended to be at the top of the doc.

lukemarsden replied:
yep

lukemarsden replied:
fixed

caseydavenport commented:
> This is a WIP and the instructions do not yet work.

D'oh, I should read things better.

luxas commented on CNAME:
```diff
@@ -1,2 +0,0 @@
-kubernetes.io
```
I don't think you meant to change this

lukemarsden replied:
Will revert before I put this up for final review

luxas commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,37 @@
+
+## Networking
+
+* [Weave Net](https://github.com/weaveworks/weave-kube) is a fast, reliable pod network that carries on working even in the face of network partitions, and doesn't have any infrastructure or database dependencies. Install it on a cluster whose kubelets are configured to expect a [CNI network plugin](/docs/admin/network-plugins/):
```
Yes, I think that we can be a little less verbose here and instead document all addons listed in `cluster/addons` plus dashboard, ingress, etc.

cdrage replied:
I'm assuming that alternative networking would be documented here as well in the future? (ex. when Canal combines weave + flannel or 1.0.0 of Flannel)

caseydavenport replied:
@cdrage Yes, I have a PR open against this branch to do that very thing!

https://github.com/lukemarsden/lukemarsden.github.io/pull/1

lukemarsden replied:
+1

I think we do need a 1 sentence description of each add-on though otherwise users have to click all the links to know what they're looking at. Keeping it to 1 sentence and linking to more verbose docs at the appropriate URL totally makes sense though.

lukemarsden replied:
I also think eventually a table view might be good for this. It'll force us to keep the one sentence descriptions short. I would like to keep the `kubectl apply -f` copyable commands though to demonstrate how easy it is to install these add-ons to users.

caseydavenport replied:
I really don't think the `kubectl apply -f` commands should be documented here.  This needs to be just a collection of links, with any documentation living elsewhere.

A _short_ description (less than one line) is fine, but that should be it.  Installation instructions should live with the addon documentation.

lukemarsden replied:
OK, I've removed the `kubectl apply -f` instructions from the addons page.

luxas commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,37 @@
+
+* [Weave Net Policy](https://github.com/weaveworks/weave-npc/tree/initial-implementation) extends Weave Net to support the [Kubernetes policy API](/docs/user-guide/networkpolicies/) so that you can securely isolate different pods from each other based on namespaces and labels:
+
+      $ kubectl apply -f https://raw.githubusercontent.com/weaveworks/weave-npc/40f5461f2f840eb8a223710e227f687cbaa55d0f/k8s/daemonset.yaml
```
This isn't great. If you strongly want this, tag commit weaveworks/weave-npc@40f5461f2f840eb8a223710e227f687cbaa55d0f with a version like `v0.1.0` or `v1.0.1`

jbeda replied:
I'd love to see instructions like this move to a weave owned page.  It'll be easier to keep up to date as weave changes and will scale better as we document more types of add-ons.

If we have community supported add-ons (that are part of the k8s project) perhaps we should create new pages for those so that they keep this page clean?

lukemarsden replied:
Yeah, this was just a temporary placeholder URL. `weave-npc` is going to be merged into `weave-kube` so I will kill this section entirely.

lukemarsden replied:
fixed

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+
+You will:
+
+1. Install packages on all machines
```
Install the `kubeadm` package on all machines

jbeda replied:
Mention prerequisites?  You'll need to have docker installed on each machine starting out.

cdrage replied:
install what exactly? (kubectl / docker / rkt / kubeadm?)

lukemarsden replied:
This tl;dr is not meant to be followed, it's meant to give a sense of how easy the process is to encourage people to continue. Prerequisites are mentioned below.

lukemarsden replied:
Also, you don't need to have docker installed on each machine before starting out.

lukemarsden replied:
I've killed the tl;dr section, it was basically a repetition of the objectives section and was causing confusion.

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+1. Run `kubeadm init` on the machine you want to become a master
+1. Run `kubeadm join` on the machines you want to become nodes
+
+You will then be able to install a pod network with `kubectl apply`.
```
...install a CNI overlay network with `kubectl apply`

caseydavenport replied:
I'd prefer _not_ to use "CNI overlay network"

If anything, call it a "CNI network provider" or "CNI network plugin"  - not all of these are overlays!

lukemarsden replied:
I didn't want to mention CNI because users shouldn't have to understand what that is to use this. I can call it a "network provider" rather than "pod network" if there's a strong feeling about this, but I like "pod network" because users probably grok pods, and grok that they need a network...

lukemarsden replied:
I've gone with pod network everywhere in this doc for consistency and also described why you need one: "You must install a pod network add-on so that your pods can communicate with each other when they are on different hosts."

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+## Prerequisites
+
+1. One or more machines running Ubuntu 16.04 or Fedora 24
+1. 2GB or more of RAM per machine
```
Where does this come from?

cdrage replied:
Pr-requisites should be higher up ^^

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+
+1. One or more machines running Ubuntu 16.04 or Fedora 24
+1. 2GB or more of RAM per machine
+1. A network connection between the machines with all ports open
```
This sentence is really scary. 

To me it sounds like:
_We're going to install a secure k8s cluster, let's open all ports on all machines!_

caseydavenport replied:
+1 - It should be clear which ports are required.

e.g "The machines will need to be able to reach the master on TCP X, and other nodes on TCP Y,Z"

jbeda replied:
This is tricky.  By default most folks will run wide open between nodes in the cluster but they may want to lock it down.  Perhaps say "full connectivity" instead of "all ports open".  We can then have an advanced section that talks about how to run over wider links or in a more locked down way?

luxas replied:
The default setup will be quite locked down and secure, so I think saying `full connectivity between all machines` like @jbeda suggested is best

But yes, we should have more in-depth docs somewhere also...

cdrage replied:
what about private networking? seems better to say "full connectivity is required" and something about preferring private networking.

lukemarsden replied:
I'm happy to write an in-depth doc that describe more precise, detailed requirements and deployment options and also explain the nitty gritty of what `kubeadm` is actually doing. But I'll do that in a second pass. For now I'll clarify that this only means that there should be full connectivity _between the machines_ not between the machines and the internet. And I guess we can mention that private networks also work :)

lukemarsden replied:
Does anyone have a list of the TCP ports required to be open between nodes/masters? If so I will add the here. We can easily amend this doc after the release as well btw.

lukemarsden replied:
I've got with "A network connection with open ports between the machines (public or private network is fine)"

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+## Objectives
+
+1. Install a secure Kubernetes cluster on your machines
+1. Install a pod network on the cluster so that application components (pods) can talk to eachother
```
each other

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+1. Install a secure Kubernetes cluster on your machines
+1. Install a pod network on the cluster so that application components (pods) can talk to eachother
+1. Install a sample microservices application (a socks shop) on the cluster
+
```
Probably a non-objectives would fit in here also

lukemarsden replied:
I don't think that's necessary.

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+* If the machine is running Ubuntu 16.04, run:
+
+      # curl -sSL https://get.docker.com/ | sh
+      # apt-get install -y socat
```
The real flow here will be 
```
curl -sSL https://get.docker.com/ | sh
echo "deb https://packages.cloud.google.com/apt kubernetes-xenial-unstable main" > /etc/apt/sources.list.d/k8s.list
curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -
apt-get update && apt-get install -y kubeadm
```

@mikedanese We might want to promote kubeadm from unstable on the real release

luxas replied:
Also point out that the kubeadm package will install kubelet, kubectl and kubernetes-cni as well, and other dependencies of the kubelet

lukemarsden replied:
That's not quite right (we're using Xenial's docker now), but yes, I'll update this once we have official debs.

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+      enabled=1
+      gpgcheck=0
+      EOF
+      # yum install kubelet kubeadm kubectl kubelet-plugin-cni
```
Only `kubeadm` should be specified here, kubeadm should depend on the rest.
Also I think `kubelet-plugin-cni` should be named `kubernetes-cni` as on the debian side
@dgoodwin ^

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+      # dpkg -i debian/bin/*.deb
+
+   If the machine is running Fedora 24, run:
+
```
First docker should be installed from `curl -sSL https://get.docker.com/ | sh`

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+The master is the machine where the "control plane" components run, including `etcd` (the cluster database) and the API server (which the `kubectl` CLI communicates with).
+All of these components will run in containers started by `kubelet`.
+
+To initialize the master, pick one of the machines you previously installed `kubelet` and `kubeadm` on, and run:
```
/`kubelet` and `kubeadm`/`kubeadm`/

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+
+      # kubeadm init --schedule-workload
+
+* If you do not want to be able to schedule workloads on the master (perhaps for security reasons), run:
```
Switch these two, `kubeadm init` is more important and should be described first

lukemarsden replied:
I wanted to put the use-case for development up first, to force people to think about whether they want that or not. I think it's likely to be more common than the security-conscious `--schedule-workload=false`.

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+    # kubeadm join --token f0c861.753c505740ecde4c 138.68.135.192
+
+A few seconds later, you should notice that running `kubectl get nodes` on the master shows a cluster with as many machines as you created.
+Your cluster is now bootstrapped!
```
securely bootstrapped

luxas commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,210 @@
+
+## Limitations
+
+1. The cluster we create here won't have cloud-provider integrations, so for example won't work with (for example) [Load Balancers](/docs/user-guide/load-balancer/) (LBs) or [Persistent Volumes](/docs/user-guide/persistent-volumes/walkthrough/) (PVs).
```
There is the cloud-provider option, we should document that if the user specifies that, it will (or should) work

stp-ip replied:
LBs and PVs are possible, just not out of the box. Wouldn't it be enough to say, that the above guide does have some limitations especially within automatic and convenient solutions for LBs and PVs?

lukemarsden replied:
I've never seen this work, and I don't know what you have to do to your cloud provider instances to make it possible for them to provision PVs and LBs. If someone can educate me on that, though, I'd be happy to document it :)

luxas commented:
@kubernetes/sig-cluster-lifecycle

dgoodwin commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,212 @@
+      # # XXX Shouldn't be necessary, need to add this to the kubeadm configure step
+      # systemctl daemon-reload && systemctl restart kubelet
+
+   If the machine is running Fedora 24, run:
```
Also should be valid for CentOS 7, and Red Hat Enterprise Linux 7.

dgoodwin replied:
Hmm, except on Fedora technically it should be "dnf" not "yum" for the commmand invocations. I think yum is still aliased to dnf though so they should work there. (/etc/yum.repos.d is valid in either case)

lukemarsden replied:
we're switching to CentOS 7 now, will update accordingly

lukemarsden commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,37 @@
+---
```
I am thinking what makes most sense is to split the definition of add-ons into "built-in add-ons" and "3rd party add-ons". The former will obey the semantics in the [cluster addons README](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons) and the latter will be "just `kubectl apply -f` it" style external add-ons like Weave Net. How does that sound @caseydavenport?

caseydavenport replied:
Hmm, I think there is some overlap between this comment and the comments below.

The best thing in the short term is probably to:

- keep calling all of these things addons, but not mandating an installation method.  Each addon should document its own installation procedure.  This keeps the docs leaner and easier to maintain, since the documentation for each addon _lives with that addon_, and some addons will be more / less complex than others.

- We can link to each of the existing `/cluster/addons/X` addons in this doc.

- As the old `/cluster` directory is torn down, those existing addons will either find new homes or go away.

WDPT?

jbeda replied:
I'd love to see all addons move to being `kubectl apply -f`.  It'll be much easier to install and manage.  They were done the way they were done because we didn't have daemonsets at the time, etc.

caseydavenport replied:
See my comment here: https://github.com/kubernetes/kubernetes.github.io/pull/1265/files#r80270865

To summarize, I think the clarification re: addons vs the "/cluster/addons" directory is much better!  I still think we could improve the rest of the page so that it comes across less like marketing.

philips commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,212 @@
+
+You can learn more about Weave for Kubernetes on the project's [GitHub page](https://github.com/weaveworks/weave-kube).
+
+You can see a complete list of available network add-ons on the [add-ons page](/docs/admin/addons/).
```
+1 I love all of the work y'all are doing. But please lets centralize the docs to a single list and then link off to the external docs for installation. These docs are going to sprawl and break if we don't focus it to a single location.

lukemarsden replied:
Sure, will make this change.

lukemarsden replied:
I've moved the "You can see a complete list of available network add-ons" line up, and then repeated it again below.

caseydavenport commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,37 @@
+
+This page lists some of the available add-ons.
+
+## Networking
```
I've verified that the following networking addons work with kubeadm clusters:

- Calico: https://github.com/projectcalico/calico-containers/tree/master/docs/cni/kubernetes/manifests/kubeadm
- Canal: https://github.com/tigera/canal/tree/master/k8s-install/kubeadm

@lukemarsden how would you like to add these?  Should I just open a PR against your branch?

lukemarsden replied:
Cool yeah please do!

jbeda commented on _data/guides.yml:
```diff
@@ -10,8 +10,10 @@ toc:
   - title: Downloading or Building Kubernetes
     path: /docs/getting-started-guides/binary_release/
-  - title: Hello World Walkthrough
+  - title: Hello World Walkthrough on GKE
```
Spell out "Google Container Engine"?

lukemarsden replied:
The title gets a bit long, but maybe "Hello World on Google Container Engine" will fit

jbeda commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,37 @@
+
+## Visualization &amp; Control
+
+* [Weave Scope](https://www.weave.works/documentation/scope-latest-installing/#k8s) is a tool for graphically visualizing your containers, pods, services etc.
```
I think a simple link here is all we should do.  Having full installation instructions for each add-on just won't scale as more vendors get involved.

stp-ip replied:
I agree with @jbeda.
Either link to a sub document (they should be responsible for) or link to documentation they own specifically. Keeping the install and instructions in the k8s repo simplifies community updates from within k8s, but it might be reasonable to let the vendor take care of this outside k8s. This would push issues and updates to the docs further to the vendor.

lukemarsden replied:
That's cool I'll move the verbose instructions over to a README on our github. Thanks!

jbeda commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+
+
+* SSH into the machine and become `root` if you are not already (for example, run `sudo su -`).
+* If the machine is running Ubuntu 16.04, run:
```
Do we want to have even more "raw" instructions on how to do this without a package manager?  If we could put it inside an &lt;details> tag?

<details>
<summary>This is an example of a details tag</summary>
Lots of *long* text here on how to do something.  We'd need to test to see if this works with the rendering pipeline.  Browser support is pretty good except for IE.  See http://www.w3schools.com/tags/tag_details.asp
</details>

lukemarsden replied:
I'd put this in the more detailed doc I'm going to write later if that's ok :)

lukemarsden replied:
Also in the more detailed doc we can explain how things work with upstream dockerproject.org docker packages but that you're on your own wrt version upgrades if you go that route.

jbeda commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+* If the machine is running Ubuntu 16.04, run:
+
+      # curl -sSL https://get.docker.com/ | sh
+      # apt-get install -y socat
```
`socat` is a strange enough dependency folks may be curious.  If we really need it why isn't it a package dependency?

luxas replied:
It is. All dependencies of kubelet will be installed automatically. This should definitely don't be here, as I pointed out above.

jbeda commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+You must install a pod network add-on so that your pods can communicate with eachother when they are on different hosts.
+
+Several projects provide Kubernetes pod networks.
+A simple one with no infrastructure or database dependencies is Weave Net, which you can install by running, on the master:
```
Can we structure this as a list so that we can put other options here?  It'd be great to have 2 supported options out the gate.

caseydavenport replied:
I can include this in my PR to add Calico / Canal to the addons page (mentioned in another comment).

cdrage replied:
`run by applying to the master` perhaps?

lukemarsden replied:
Based on user testing of this doc so far I think it's important to have one recommended option that is marked "easy to install", and I'd selfishly like that to be Weave Net. However it's also important to emphasize that the user has choice, so I will add an explicit call-out to Calico and Canal in the text. @caseydavenport if you can give me some words to use to describe them and the distinction between them and somewhere to link to, I'll add that here. Thanks!

lukemarsden replied:
I think "installing" and "running" are more commonly understood verbs than "applying" for the target audience of this doc.

caseydavenport replied:
@lukemarsden, I think we don't want this doc saying that any one solution is "recommended" by Kubernetes.  I think @jbeda's suggestion of structuring this as a list or something similar is appropriate, so long as it's not really verbose.

I've had a go at tweaking this to be agnostic, while maintaining ease of use for the guide follower: https://github.com/caseydavenport/caseydavenport.github.io/pull/1/files

What do people think of something like this?

jbeda commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+
+**You should run this on the master before you try to deploy any applications to your cluster.**
+
+Once the command has completed, a few seconds later you should see the `weave-net` pods and the `kube-dns` pod go into `Running` in the output of `kubectl get pods --all-namespaces`. **This signifies that your cluster is ready.**
```
This assumes that folks are using `weave-net`.  We need to avoid `kubeadm` looking like a weave only solution.

lukemarsden replied:
Ack.

jbeda commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+
+Here you will install the NodePort version of the Socks Shop, which doesn't depend on Load Balancer integration, since our cluster doesn't have that:
+
+    # kubectl apply -f https://raw.githubusercontent.com/lukemarsden/microservices-demo/master/deploy/kubernetes/definitions/wholeWeaveDemo-NodePort.yaml
```
Does socks shop have weave specific stuff in it?  We should make sure that this doesn't look to be tied to the networking solution.

luxas replied:
Maybe we can put it in kubernetes/kubernetes/examples for the time being?
I think the examples are on their way out from the main repo, but until the Google guys have made the kubernetes/examples repo, @lukemarsden can maintain it in the main repo

That would give a much more professional look to it.

lukemarsden replied:
The socks shop is not weave-specific, it's a generic microservices demo that will work on any kubernetes cluster (and lots of other things too) and lives in a [vendor-independent org on github](https://github.com/microservices-demo). I need to make a PR to add a NodePort version that will work with this guide and then I can replace the `lukemarsden` URL with one that links there instead.

jbeda commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+* Mailing List: [kubernetes-sig-cluster-lifecycle](https://groups.google.com/forum/#!forum/kubernetes-sig-cluster-lifecycle)
+* [GitHub Issues](https://github.com/kubernetes/kubernetes/issues): please tag `kubeadm` issues with `@kubernetes/sig-cluster-lifecycle`
+
+## Limitations
```
We should make it clear that `kubeadm` is a work in progress and that these will be addressed in the fullness of time.

lukemarsden replied:
Absolutely. Will add text to that effect here. Thanks!

lukemarsden replied:
done

mikedanese commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+    <master/apiclient> all control plane components are healthy after 61.346626 seconds
+    <master/apiclient> waiting for at least one node to register and become ready
+    <master/apiclient> first node is ready after 4.506807 seconds
+    <master/discovery> created essential addon: kube-discovery
```
is kube-discovery an addon or a static pod?

luxas replied:
It's a Deployment

lukemarsden replied:
this is going away as soon as kube-discovery gets merged into the apiserver, so I don't think we need to think too hard about this line of output right now

stp-ip commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+
+1. Install packages on all machines
+1. Run `kubeadm init` on the machine you want to become a master
+1. Run `kubeadm join` on the machines you want to become nodes
```
Should we assume everyone doing this non-expert guide to know the difference between master and nodes? Might be clearer to name them worker nodes. (The master is some kind of node too)

cdrage replied:
s,machine,machine(s),g

luxas replied:
We've already debated about this a lot.
master/node is the current convention

lukemarsden replied:
I will explain what masters and nodes are in a short sentence here, as I have tried to do with other new concepts introduced in this doc.

stp-ip commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+
+   Instead we will use the [NodePort feature of services](/docs/user-guide/services/) to demonstrate exposing the sample application on the internet.
+   To easily obtain a cluster which works with LBs and PVs Kubernetes, try [the "hello world" GKE tutorial](/docs/hellonode) or [one of the other cloud-specific installation tutorials](/docs/getting-started-guides/).
+1. The cluster we create here will have a single master, with a single `etcd` database running on it.
```
Maybe explain or link to the limitations. When does data loss occur, when is HA needed.

lukemarsden replied:
I'll add a couple words around that here, sure.

webwurst commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,37 @@
+* [Weave Scope](https://www.weave.works/documentation/scope-latest-installing/#k8s) is a tool for graphically visualizing your containers, pods, services etc.
+  Register for a [Weave Cloud account](https://cloud.weave.works/) to get a service token and then replace `<token>` with your token below:
+
+      $ kubectl apply -f https://cloud.weave.works/launch/k8s/weavescope.yaml?service-token=<token>
```
`https://cloud.weave.works/launch/k8s/weavescope.yaml?service-token=<token>` should be in quotation marks.

lukemarsden replied:
good spot!

sttts commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+
+## Overview
+
+This quickstart will show you how to easily install a secure Kubernetes cluster on any machines running Linux, using a tool called `kubeadm` which is part of Kubernetes.
```
nit: "running Linux" is not "running Ubuntu 16.04 or Fedora 24"

cdrage replied:
perhaps go into detail (any requirements for kubeadm?ex. any machine with kernel version blah blah running docker?)

lukemarsden replied:
I will wordsmith this. I want to keep it simple at the top and add complexity gradually to make for a consumable doc, without oversimplifying.

cdrage commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,249 @@
+* SSH into the machine and become `root` if you are not already (for example, run `sudo su -`).
+* If the machine is running Ubuntu 16.04, run:
+
+      # curl -sSL https://get.docker.com/ | sh
```
if we're running `curl -sSL https://get.docker.com/ | sh` to get docker, we may as well get the kubeadm binary from the github release pages...

cdrage commented on docs/kubeadm.md:
```diff
@@ -68,13 +68,10 @@ For each host in turn:
+      # apt-get install -y docker.io socat
       # curl -s -L \
-        "https://www.dropbox.com/s/shhs46bzhex7dxo/debs-9b4337.txz?dl=1" | tar xJv
+        "https://www.dropbox.com/s/tso6dc7b94ch2sk/debs-5ab576.txz?dl=1" | tar xJv
```
I don't think a dropbox link should be here...

lukemarsden replied:
Yes, this will be replaced by the packages that @mikedanese is going to bake as soon as https://github.com/kubernetes/kubernetes/pull/33262 and https://github.com/kubernetes/kubernetes/pull/32203 land. Same for RPMs, they will be hosted on packages.cloud.google.com.

lukemarsden commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+   Instead we will use the [NodePort feature of services](/docs/user-guide/services/) to demonstrate exposing the sample application on the internet.
+   To easily obtain a cluster which works with LBs and PVs Kubernetes, try [the "hello world" GKE tutorial](/docs/hellonode) or [one of the other cloud-specific installation tutorials](/docs/getting-started-guides/).
+1. The cluster we create here will have a single master, with a single `etcd` database running on it.
+   Adding HA support (multiple `etcd` servers, multiple API servers, etc) is still a work-in-progress.
```
I will add that this is a WIP only for `kubeadm` and that it can be made to work for k8s in general already.

lukemarsden commented:
> A limitation right now is that kubectl logs doesn't work I think (since the node names aren't resolvable)

I've added this to the limitations section.

lukemarsden commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,243 @@
+    # kubectl apply -f https://git.io/weave-kube
+    daemonset "weave-net" created
+
+If you prefer [Calico](https://github.com/projectcalico/calico-containers/tree/master/docs/cni/kubernetes/manifests/kubeadm) or [Canal](https://github.com/tigera/canal/tree/master/k8s-install/kubeadm), please refer to their respective installation guides.
```
@caseydavenport I have added an explicit shout out to Calico and Canal here and linked to their install instructions. If you want me to add a one-liner install command here as well, please let me know which one to use. I don't want to have more than two options in this guide because it starts to become too much cognitive overhead for the user. This is meant to be a quickstart after all :)

caseydavenport replied:
Yep, thanks for the shout out :)

My idea here was essentially that there's a bunch of thinking in the community going on as to how we can keep Kubernetes to be more vendor neutral.  Obviously I'd selfishly like Calico or Canal to be more featured in this section, but I don't think providing instructions for a lot of projects here is the right thing to do.

caseydavenport replied:
> This is meant to be a quickstart after all :)

Should this instead be in docs/getting-started-guides?

lukemarsden replied:
OK, thanks. I'll move this to `docs/getting-started-guides`.

lukemarsden commented:
@luxas @jbeda @cdrage @sttts @webwurst @mikedanese @caseydavenport @stp-ip @dgoodwin 

Thanks for all the review comments! I think I've addressed all of them now, and left one comment for you @caseydavenport.

@jaredbhatti this is really close now. The only item we needed before we can release these docs is, IMO:

[ ] https://github.com/kubernetes/kubernetes/pull/33262 merged and @mikedanese's debs and rpms released, and these docs updated to reflect the location of the official debs and rpms

I am very much hoping that we'll be able to get this done today (Friday) so I can make that one change here, test end-to-end and then we'll be in good shape to release these docs on Monday!

caseydavenport commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,24 @@
+
+This page lists some of the available add-ons and links to their respective installation instructions.
+
+## Networking and Network Policy
```
Honestly, reading this now all three of these still just feel like marketing speak, and I really don't want this to be a marketing page - especially as more solutions get added to the list.  I think this has a lot of potential to become a "my technology is better than your technology" fest.  Still feel pretty strongly that the one sentence descriptions _do not belong here_.

On a secondary note, I had thought lists like this should be in alphabetical order. I understand that puts Calico at the top, but the intention is just so that some level of order is maintained rather than everyone scrambling for the top.  An alternative that puts Weave at the top would be to use reverse alphabetical order :)

lukemarsden replied:
I'd be happy to try and make the sentences feel less like marketing speak, but I feel strongly that giving the user a description of what each thing is makes for a much better UX than forcing them to click through to each page, which would be the case if it was a bare list of project names. (Remember, to a new user, the projects' names themselves will likely be meaningless.)

caseydavenport replied:
Yeah, my feelings are that there is scope for "cookie-cutter" type statements about each project that give users a 1000ft view of what each does:
- X provides networking and network policy.
- Y provides networking.
- Z provides network policy.

Things like "simple, scalable, secure", "easy, fast and reliable", etc certainly rub me the wrong way (I understand that I provided some of that verbiage...), and listing the various features of each project is either going to be super verbose, or provide only a limited picture of each to the reader. 

If I were a user picking a networking option, this is not the page I'd use to make my decision, though it might be where I start looking for names to investigate.

I understand we might just be diametrically opposed here, and I don't want to be the last one holding up these docs.  Does anyone else have an opinion here?

lukemarsden replied:
I've toned down the 'marketing speak' in https://github.com/kubernetes/kubernetes.github.io/pull/1265/commits/b2458c467b6f1b65e689912600502b1d0f6d89b8 😄

devin-donnelly commented on _data/guides.yml:
```diff
@@ -8,10 +8,12 @@ toc:
   section:
   - title: What is Kubernetes?
     path: /docs/whatisk8s/
+  - title: Installing Kubernetes Easily on Linux
```
"Easily" verges on marketing speak and is generally not used in technical documentation. "Installing Kubernetes on Linux with kubeadm" is probably better.

devin-donnelly commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,24 @@
+
+## Overview
+
+Add-ons extend the functionality of Kubernetes in a pluggable way.
```
"Pluggable" is a weird word. Industry term of art?

Maybe "Add-ons are plug-in modules that extend Kubernetes' functionality."

devin-donnelly commented on docs/index.md:
```diff
@@ -77,13 +77,18 @@ h2, h3, h4 {
+    <a href="/docs/hellonode/" class="button">Get Started on GKE</a>
+  </div>
+  <div class="col3rd">
+    <h3>Installing Kubernetes Easily on Linux</h3>
```
I'd still strike "Easily". Maybe "Installing Kubernetes on Linxu with kubeadm".

devin-donnelly commented on docs/index.md:
```diff
@@ -77,13 +77,18 @@ h2, h3, h4 {
+  <div class="col3rd">
+    <h3>Installing Kubernetes Easily on Linux</h3>
+    <p>This quickstart will show you how to install a secure Kubernetes cluster on any computers running Linux, using a tool called <code>kubeadm</code> which is part of Kubernetes. It'll work with local VMs, physical servers and/or cloud servers, either manually or as part of your own automation. It is currently in alpha but please try it out and give us feedback!</p>
+    <a href="/docs/kubeadm/" class="button">Install Kubernetes Easily</a>
```
"Install Kubernetes with kubeadm"

devin-donnelly commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,243 @@
+
+## Overview
+
+This quickstart shows you how to easily install a secure Kubernetes cluster on machines running Ubuntu 16.04 or CentOS 7, using a tool called `kubeadm` which is part of Kubernetes.
```
Run-on sentence.

"This quickstart shows you how to install a secure Kubernetes cluster on machines running Ubuntu 16.04 or CentOS 7. The installation uses a tool called `kubeadm`, which is part of Kubernetes 1.4."

devin-donnelly commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,243 @@
+
+This quickstart shows you how to easily install a secure Kubernetes cluster on machines running Ubuntu 16.04 or CentOS 7, using a tool called `kubeadm` which is part of Kubernetes.
+
+This process should work with local VMs, physical servers and/or cloud servers.
```
Remove "should". "This process works with..."

devin-donnelly commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,243 @@
+This quickstart shows you how to easily install a secure Kubernetes cluster on machines running Ubuntu 16.04 or CentOS 7, using a tool called `kubeadm` which is part of Kubernetes.
+
+This process should work with local VMs, physical servers and/or cloud servers.
+It is intended to be simple enough that you can easily integrate its use into your own automation (Terraform, Chef, Puppet, etc).
```
"It is intended" -> "It is"

caseydavenport commented on docs/kubeadm.md:
```diff
@@ -0,0 +1,243 @@
+
+## Cleanup
+
+* To uninstall the socks shop, run `kubectl delete -f https://raw.githubusercontent.com/lukemarsden/microservices-demo/master/deploy/kubernetes/definitions/wholeWeaveDemo-NodePort.yaml`.
```
Is this link correct?  Shouldn't this be:

```
kubectl delete -f microservices-demo/deploy/kubernetes/manifests
```

lukemarsden replied:
Yep, good shout!

spiffxp commented:
@spiffxp note to self, this set of feature issue docs for 1.4 isn't done yet

pwittrock commented:
When this is merged you must update the CHANGELOG.md with the docs link in kubernetes/kubernetes master.  See this PR which removes the TODO:

https://github.com/kubernetes/kubernetes/pull/33418/files

simonswine commented:
Would be great if we could have some really generic instructions in the getting started (without packages for example for running on CoreOS):

This worked for a custom build of the latest HEAD of the [implementation PR](https://github.com/kubernetes/kubernetes/pull/33262):

```
# Install kubectl, kubelet and kubeadm on every host
mkdir -p /opt/bin
curl -o /opt/bin/kubeadm https://s3-eu-west-1.amazonaws.com/jetstack.io-kubernetes-builds/release/v1.5.0-alpha.0.1375.a023085a5fa018/bin/linux/amd64/kubeadm
curl -o /opt/bin/hyperkube https://s3-eu-west-1.amazonaws.com/jetstack.io-kubernetes-builds/release/v1.5.0-alpha.0.1375.a023085a5fa018/bin/linux/amd64/hyperkube
chmod +x /opt/bin/kubeadm /opt/bin/hyperkube
ln -s hyperkube /opt/bin/kubectl
ln -s hyperkube /opt/bin/kubelet
curl -L https://github.com/containernetworking/cni/releases/download/v0.3.0/cni-v0.3.0.tgz | tar xvzf - -C /opt/cni/bin/
mkdir -p /opt/cni/bin

cat > /etc/systemd/system/kubelet.service <<'EOF'
[Unit]
Description=kubelet: The Kubernetes Node Agent
Documentation=http://kubernetes.io/docs/

[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--kubeconfig=/etc/kubernetes/kubelet.conf --require-kubeconfig=true"
Environment="KUBELET_SYSTEM_PODS_ARGS=--pod-manifest-path=/etc/kubernetes/manifests --allow-privileged=true"
Environment="KUBELET_NETWORK_ARGS=--network-plugin=cni"
Environment="KUBELET_DNS_ARGS=--cluster-dns=100.64.0.2 --cluster-domain=cluster.local"
Environment="KUBELET_EXTRA_ARGS=--v=4"
ExecStart=/opt/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_SYSTEM_PODS_ARGS $KUBELET_NETWORK_ARGS $KUBELET_DNS_ARGS $KUBELET_EXTRA_ARGS

Restart=always
StartLimitInterval=0
RestartSec=10

[Install]
WantedBy=multi-user.target
EOF
```

I also liked @errordeveloper's approach with using a docker image to bootstrap kubelet/kubeadm/cni.

devin-donnelly commented:
Any word on these changes, @lukemarsden ? I'd like to get this in before tomorrow morning when we merge and launch.

lukemarsden commented:
@devin-donnelly I'll have the changes made and this PR ready to go for you before you wake up today (unless you wake up _really_ early 😄). However @mikedanese indicated to me that he is likely to be cutting the 1.4 release on Tuesday. Is that right?

lukemarsden commented:
@caseydavenport @devin-donnelly I've addressed your review feedback in my latest two commits. Thanks!

@devin-donnelly I think you can remove "Tech Review: Open Issues" now. And hopefully "Docs Review: Open Issues" as well :)

@pwittrock We fixed the changelog in https://github.com/kubernetes/kubernetes/pull/33262/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1eL231 which landed this morning.

This PR is now only blocked on @mikedanese updating packages.cloud.google.com with 1.4.0 packages that include `kubeadm`. I will update this PR and remove the WIP tag soon as this has happened.

luxas commented on docs/admin/addons.md:
```diff
@@ -0,0 +1,24 @@
+
+## Visualization &amp; Control
+
+* [Weave Scope](https://www.weave.works/documentation/scope-latest-installing/#k8s) is a tool for graphically visualizing your containers, pods, services etc. Use it in conjunction with a [Weave Cloud account](https://cloud.weave.works/) or host the UI yourself.
```
Mind adding dashboard here also? That's kind of helpful for new users

lukemarsden replied:
done, thanks

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+
+## Overview
+
+This quickstart shows you how to easily install a secure Kubernetes cluster on machines running Ubuntu 16.04 or CentOS 7.
```
How about Fedora 24 and RHEL?

lukemarsden replied:
we haven't tested them so I'm not going to advertise that we work with them. We can do so once we have tested them.

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+## Prerequisites
+
+1. One or more machines running Ubuntu 16.04 or CentOS 7
+1. 2GB or more of RAM per machine
```
Why is this here?
I'm running Kubernetes fine on my 1GB Raspberry Pi's, this only confuses users, as 2GB is not a minimum
I've also run Kubernetes on 512 MB droplets

Please remove

lukemarsden replied:
I've updated this to 1GB and mentioned that any less than that would leave very little room for a user's apps.

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+
+1. One or more machines running Ubuntu 16.04 or CentOS 7
+1. 2GB or more of RAM per machine
+1. A network connection with open ports between the machines (public or private network is fine)
```
Why not `Full connectivity between all machines in the cluster`? Maybe also add `(they should be able to ping each other)`

lukemarsden replied:
Added the former.

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+* `kubeadm`: the command to bootstrap the cluster.
+
+For each host in turn:
+
```
Mention that `apt-transport-https` is required if it doesn't exist

lukemarsden replied:
added

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+    deb http://packages.cloud.google.com/apt kubernetes-xenial-unstable main
+    EOF
+    # apt-get update
+    # apt-get install -y kubelet-kubeadm kubectl
```
`apt-get install -y kubeadm docker-engine=1.11.2`

Nothing more

luxas replied:
Or `docker.io` if we want to use Ubuntu's packages (which I think we want)

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+* SSH into the machine and become `root` if you are not already (for example, run `sudo su -`).
+* If the machine is running Ubuntu 16.04, run:
+
+      # apt-get install -y docker.io socat
```
This should be like the outcommented code above (with my correction)

lukemarsden replied:
yes, but i'm only going to change the docs after the packages are published (we have people testing the docs)

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+      enabled=1
+      gpgcheck=0
+      EOF
+      # yum install docker kubelet kubeadm kubectl kubernetes-cni
```
@dgoodwin When you get the deps right, it should be just `kubeadm`
Deps: 
kubeadm => kubelet, kubectl
kubelet => kubernetes-cni

lukemarsden replied:
I'll wait for @dgoodwin to ask me to change the instructions in the docs

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+## What's next
+
+* Learn more about [Kubernetes concepts and kubectl in Kubernetes 101](/docs/user-guide/walkthrough/).
+* Install Kubernetes with [a cloud provider configurations](/docs/getting-started-guides/) to add Load Balancer and Persistent Volume support.
```
Why is this here?

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+
+Please note: `kubeadm` is a work in progress and these limitations will be addressed in due course.
+
+1. The cluster created here doesn't have cloud-provider integrations, so for example won't work with (for example) [Load Balancers](/docs/user-guide/load-balancer/) (LBs) or [Persistent Volumes](/docs/user-guide/persistent-volumes/walkthrough/) (PVs).
```
It doesn't _by default_, but refer to the `kubeadm reference` doc if you want to do enable cloud provider integrations

lukemarsden replied:
we can add ref to that doc once it exists

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+   Workaround: use `docker logs` on the nodes where the containers are running as a workaround.
+1. There is not yet an easy way to generate a `kubeconfig` file which can be used to authenticate to the cluster remotely with `kubectl` on, for example, your workstation.
+
+   Workaround: copy the kubelet's `kubeconfig` from the master: use `scp root@<master>:/etc/kubernetes/kubelet.conf .` and then e.g. `kubectl --kubeconfig ./kubelet.conf get nodes` from your workstation.
```
`/etc/kubernetes/admin.conf` should be copied instead of kubelet.conf

lukemarsden replied:
fixed

luxas commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,245 @@
+## Cleanup
+
+* To uninstall the socks shop, run `kubectl delete -f microservices-demo/deploy/kubernetes/manifests` on the master.
+* To uninstall Kubernetes, simply delete the machines you created for this tutorial.
```
I have a script for this, which we should refer to. Saying delete your machines is kind of awful :smile:
```shell
systemctl stop kubelet; 
docker rm -f $(docker ps -q); mount | grep "/var/lib/kubelet/*" | awk '{print $3}' | xargs umount 1>/dev/null 2>/dev/null; 
rm -rf /var/lib/kubelet /etc/kubernetes /var/lib/etcd /etc/cni; 
ip link set cbr0 down; ip link del cbr0; 
ip link set cni0 down; ip link del cni0;
systemctl start kubelet; 
```

lukemarsden replied:
I'll put this inside `<details>` tag

luxas commented:
@lukemarsden Also, please squash the commits

lukemarsden commented:
@luxas I've addressed your review feedback and squashed. Thanks!

luxas commented:
The only issue left is the package installation guide, but I guess that will be cleaned up when @mikedanese have pushed the final packages

lukemarsden commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,254 @@
+* SSH into the machine and become `root` if you are not already (for example, run `sudo su -`).
+* If the machine is running Ubuntu 16.04, run:
+
+      # apt-get install -y docker.io socat apt-transport-https
```
These instructions need to be updated when @mikedanese's official packages land.

lukemarsden commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,254 @@
+
+   If the machine is running CentOS 7, run:
+
+      # cat <<EOF > /etc/yum.repos.d/k8s.repo
```
These instructions need to be updated when @mikedanese's official packages land.

lukemarsden commented on docs/getting-started-guides/kubeadm.md:
```diff
@@ -0,0 +1,254 @@
+
+To initialize the master, pick one of the machines you previously installed `kubelet` and `kubeadm` on, and run:
+
+     # kubeadm init --use-kubernetes-version v1.4.0-beta.11
```
Need to remove `--use-kubernetes-version v1.4.0-beta.11` when @mikedanese's official packages land.
